consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,209 @@
+	blockchain blockchainBackend
+	// consensusBackend is abstraction for polybft consensus specific functions
+	consensusBackend polybftBackend
+	// rootchain represents abstraction for rootchain interaction
+	rootchain rootchainInteractor
+	// checkpointsOffset represents offset between checkpoint blocks (applicable only for non-epoch ending blocks)
+	checkpointsOffset uint64
+}
+
+// newCheckpointManager creates a new instance of checkpointManager
+func newCheckpointManager(sender types.Address, blockTime time.Duration, interactor rootchainInteractor,
+	blockchain blockchainBackend, backend polybftBackend) *checkpointManager {
+	r := interactor
+	if interactor == nil {
+		r = &defaultRootchainInteractor{}
+	}
+
+	return &checkpointManager{
+		sender:            sender,
+		blockchain:        blockchain,
+		consensusBackend:  backend,
+		rootchain:         r,
+		checkpointsOffset: uint64(checkpointTimeInterval.Milliseconds() / blockTime.Milliseconds()),
+	}
+}
+
+// getCurrentCheckpointID queries CheckpointManager smart contract and retrieves current checkpoint id
+func (c checkpointManager) getCurrentCheckpointID(epochNumber uint64) (uint64, error) {
+	checkpointIDMethodEncoded, err := currentCheckpointIDMethod.Encode([]interface{}{})
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number for epoch=%d: %w",
+			currentCheckpointID, epochNumber, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, epochNumber uint64) error {
+	checkpointID, err := c.getCurrentCheckpointID(epochNumber)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNum := checkpointID + 1; blockNum < latestHeader.Number; blockNum++ {
+		header, exists := c.blockchain.GetHeaderByNumber(blockNum)
+		if !exists {
+			return fmt.Errorf("block %d was not found", blockNum)
+		}
+
+		extra, err := GetIbftExtra(header.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		// send checkpoints only for epoch ending blocks (validators delta is non-empty)
+		if extra.Validators.IsEmpty() {
```

Comment:
Is it a marker of epoch ending?

Reply:
Yes: only epoch ending blocks have non-empty validator set delta.

At first I was thinking to determine epoch ending blocks by epoch size, but it turned out that it cannot be easily fitted here, because we don't know whether `checkpointID` corresponds to the epoch-ending or some block within the epoch. Therefore I find this check more robust (plus it is going to work properly with slashing as well).